debug_reg_printer.c

```diff
@@ -46,6 +46,9 @@ static unsigned int riscv_debug_reg_field_to_s(char *buf, unsigned int offset,
 		riscv_debug_reg_field_info_t field, riscv_debug_reg_ctx_t context,
 		uint64_t field_value)
 {
+	if (field_value == 0)
```

Comment:
Wouldn't it be preferable to only skip value if it is zero and there is no name for such field value?
Like so:
```
static const char *get_field_value_name(riscv_debug_reg_field_info_t field,
                uint64_t field_value)
{
        return field.values ? field.values[field_value] : NULL;
}
```
```suggestion
	if (field_value == 0  && !get_field_value_name(list.field, field_value))
```

Reply:
I don't think that's preferable. Almost everywhere the 0 value means inactive, so not interesting to list. 
E.g. dmstatus=0 would be `{version=none stickyunavail=current ndmresetpending=false confstrptrvalid=invalid authbusy=ready authenticated=false}` which is a lot to visually parse through just to see nothing is going on. And there's nothing preventing the XML to be edited in the future to be more consistent or whatever and to name every 0 field, in which case the output would be even larger.

Reply:
> I don't think that's preferable. Almost everywhere the 0 value means inactive, so not interesting to list.

The issue, in my opinion is that it is exactly **almost** everywhere. E.g. `dtmcs.version == 0 ` means `0.11` which is quite important information, considering the use of this code in OpenOCD's `riscv-013.c`.

> E.g. dmstatus=0 would be `{version=none stickyunavail=current ndmresetpending=false confstrptrvalid=invalid authbusy=ready authenticated=false}` which is a lot to visually parse through just to see nothing is going on.

IMHO this particular issue is caused by attempting to get a meaningful result from decoding a meaningless value and should be addressed in OpenOCD code. Please, take a look at [this PR](https://github.com/riscv/riscv-openocd/pull/951).

Reply:
Ideally we could mark the handful of fields where 0 is significant, but I'm not going to do that work right now. It's much better to have the output be concise than it is to be readable by someone with no experience.

I chose the 0 example because it was easy. The problem is just as bad in a realistic example:
```dmstatus=0x430382 {version=0.13 anyresumeack=1 allresumeack=1 impebreak=1 stickyunavail=current ndmresetpending=false confstrptrvalid=invalid authbusy=ready authenticated=true anyhalted=1 allhalted=1}```
vs
```dmstatus=0x430382 {version=0.13 anyresumeack=1 allresumeack=1 impebreak=1 authenticated=true anyhalted=1 allhalted=1}```

The former has a bunch of uninteresting fields that just make it harder to read the ones you usually care about.

On top of that, the XML might change in the future to add more explicit true/false encodings, which would make the problem worse. The contents of that XML are primarily for the formatting of the document, not for this automated field generation.

Reply:
@timsifive 

> Ideally we could mark the handful of fields where 0 is significant,
> It's much better to have the output be concise than it is to be readable by someone with no experience.


The former seems an unfeasible goal since it won't be easy to maintain (you've actually alluded to that already). The latter seems like an aesthetic preference  - there are people who prefer verbose printouts and there are people who hate these logs and want them to be as short as possible (me included). However, I can see the reason some may want to decode every field. I'd like to suggest a compromise: we can add an extra argument to the printing function and a corresponding riscv-specific OpenOCD command to control it. So we can enable verbose decoding of fields. I can even volunteer to provide a patch (if no objections).

Reply:
OK. Done.